### ISSUE TITLE
Create multiple plates in pending state at once

### ISSUE THREAD

**Issue by yoldas** [sanger/limber#2076]
**User story**

As Limber user running LCM Triomics EMSeq and WGS pipeline, I would like to create multiple successive plates in pending state at once and print barcodes.

"To generate the LCMT DNA Frag, LCMT DNA End Prep, LCMT DNA Adp Lig, LCMT TET2 Ox & LCMT TET2 Stop plates at the beginning of the process required me to serially ‘Add an empty LCMT X plate’ on each parent plate to create each child plate in turn, and separately print barcodes each time."

**Who are the primary contacts for this story**

Scott G.
Abdullah

**Who is the nominated tester for UAT**

Scott G.

**Acceptance criteria**

To be considered successful the solution must allow:

Option 1) A create button on a parent plate **triggers** the labware creation immediately and shows the child plates in a custom page and allows printing barcodes. (create and show actions).

- [ ] There is a button with a configured text on the sidebar to open a custom page to create successive child plates in pending state. For example, such a button (with the text Bravo Initial Setup) for the setup will be available on the LCMT Lysate plate to create LCMT DNA Frag, LCMT DNA End Prep, LCMT DNA Adp Lig, LCMT TET2 Ox & LCMT TET2 Stop child plates, and also another (with the text Bravo LCMT EM TET2 Stop to Denat and Deam Setup) on LCMT EM TET2 Stop to create LCMT EM NaOH Denat and LCMT EM APOBEC Deam child plates.
- [ ] The custom page creates the successive child plates, list the barcodes, and gives an option to print barcodes.
- [ ] The custom page is backed by a specialised labware creator to create and link successive child plates of configured purposes.
- [ ] The custom page lists the children barcodes, and a form to print all barcodes at once.
- [ ] The custom page has links to the individual labware pages.
- [ ] The purpose config of the parent plate contains a section for the specialised labware creator, and child purposes and button text as arguments.
- [ ] The parent plate presenter renders an additional section in the sidebar for the feature and shows the button.

Option 2) A create button on a parent plate opens an **intermediate** page to present the configuration for the child creation and a Create button. This intermediate page corresponds to new action of the controller. Once the button is clicked, it presents another page with the same functionality in option 1 (create and show actions). Having an intermediate page is to present the description of the action and configuration before the labware creation.

Once the feature is implemented:
- [ ] Use it to create LCMT DNA Frag, LCMT DNA End Prep, LCMT DNA Adp Lig, LCMT TET2 Ox & LCMT TET2 Stop from  LCMT Lysate
- [ ] Use it to create LCMT EM NaOH Denat and LCMT EM APOBEC Deam from LCMT EM TET2 Stop

**Dependencies**

This story is blocked by the following dependencies:
N/A

References
This story has a non-blocking relationship with:
- #<issue_no.>
- sanger/limber#4288

**Additional context**

This is a generic feature which may be useful to create multiple child labware at once but but it aims to solve a specific issue in the UX of the LCM Triomics EMSeq and WGS pipeline because the existing functionality makes the user jump from page to page to create new plates and print barcodes which is time consuming and error prone.

**Comment by KatyTaylor** [comment]
I like the ideas. Should be useful wherever we are currently using Permissive Presenters - as a replacement for them maybe, or an alternative.

There's some duplication in the acceptance criteria for Option 1).

**Comment by yoldas** [comment]
@KatyTaylor I have removed the duplicate criteria item for option 1).

This story also assumes that the plates to be created do not have a custom creation page to collect additional input for the child plates but such a feature could be implemented as part of option 2).

**Comment by KatyTaylor** [comment]
> This story also assumes that the plates to be created do not have a custom creation page to collect additional input for the child plates but such a feature could be implemented as part of option 2).

That sounds complicated, I wonder if any of our existing labware with permissive presenters have custom creation pages... Feels like often they would require the transfer to have happened already.

**Comment by yoldas** [comment]
This story assumes that the plates to be created do not have a custom creation page. I have just mentioned how collecting additional input would be possible but this story does not aim that.

**Comment by seenanair** [comment]
Following are the sequence diagrams and UI mockups created two different options

Option 1 - Displaying an in-place panel for multiple plate creation in 'Suggested Actions" in 'LCMT Lysate' page
https://lucid.app/lucidchart/7e44bb22-bb3a-4995-ac59-f11b3d147f6d/edit?invitationId=inv_a26e29c2-a9a2-4ee2-9994-aba6c4239d5f&page=0_0#

Option - Creatlng a link to the multiple plate creation page in 'Suggested Actions' for in 'LCMT Lysate' page
https://lucid.app/lucidchart/c6efc655-d2a2-49d6-9aaf-426a1a10bf36/edit?invitationId=inv_2f5c2396-2861-4de4-bd08-b91398289ac1&page=0_0#

**Comment by KatyTaylor** [comment]
**Summary of discussion 19/2/25 - Seena, Andrew, Abdullah, Katy**

* We put this back into the backlog due to uncertainties about:
    * Its priority, given they do have a current solution. What is the volume of samples going through this pipeline?
    * How much effort we should put into making this feature generic. How many existing pipelines would benefit from this? Is this design (long robot method with many plates loaded upfront into a carousel) going to be common in pipelines going forward?
* We decided to arrange a meeting with R&D to discuss the above questions (Katy action) - after which we can reassess this story's priority.
* We also discussed the ideas of:
    * PSD being more involved in the design of automation for pipelines / software being taken into account
    * Potential knowledge sharing sessions with R&D
        * How the software apps work / what kind of thing is hard to develop
        * Something on how they design automation - process they follow, directions they're moving in

**Comment by KatyTaylor** [comment]
**Summary of discussion with Scott & Scott (R&D) 27/2/25 (emailed as well):**

**TL;DR**
 
I’ve put this story back in the backlog for now, and will update Sujit with the relevant information on priority. Do get in touch with him if anything changes the priority, or you think it’s been forgotten about!
 

**1. How important is it to do this story from the point of view of LCM Triomics?**
 
* The pipeline is operational and has been run several times, with the functionality as-is. The users are aware of the current process (clicking through the multiple screens of ‘pending’ plates in order to print barcodes) and are happy to do it for now.
* Throughput is currently only a couple of plates a month – the pipeline wasn’t up and running in time for when a large load of samples went through. There should be more demand in future but can’t predict how much. Adding the RNA and WGS-only branches should increase throughput.
* Therefore, it is not high priority at the moment with regard to this pipeline.
* However, it would be a good user experience improvement, and the users are expecting it to arrive at some point.
 
**2. How important is this as a generic feature for Limber? Is this design a common one for R&D and is it the direction they’re moving in?**
 
* There is an ambition in future to move towards a more modular approach, from an automation point of view – instead of having discrete pipelines running in parallel doing very similar things, having shorter, reusable robot methods used in several pipelines. So, the opposite to the approach here, where a long robot method is used that does many steps in a row, with very little hands on time from the lab staff. However, this change is a long way off as needs a lot of work (on software, automation, process, scheduling).
* Even when the above is implemented, there would probably still be the occasional need for this feature in particular situations.
* It would have been a useful feature in other pipelines. For instance, the current RNA pipelines also have places where they step through multiple plates (not as many) to print the barcodes in advance. The LCMB pipeline was designed in an alternative way where the intermediate plates were not tracked – but they prefer them to be tracked, and it is necessary for the LCM Triomics pipeline because there are multiple points where the plates get taken on and off deck, and users can get lost as to which one they’re currently on (robot software can help but LIMS is preferred).
* Therefore, it’s not going to be a central/core feature, used for every pipeline, but it will bring user experience improvements to current pipelines, and is likely to continue to be a useful feature in the future.
 
**3. More detailed/technical discussion about how this feature would work**
 
* Not worried about losing the audit trail of when plates get created in the LIMS matching when they get transferred into in the lab – they don’t report on this data currently, and if they did, they could use different data e.g. the transfers/state changes.
* Mistakes part way through – would just have to repeat the whole step from the beginning – couldn’t think of an example when it would be OK to restart from part way through.
* Concerns about leaving abandoned plates ‘in limbo’ if the whole step was repeated – could check the RNA pipeline to see if this is currently an issue. If so, could think about a bulk cancellation feature or something.
* Starting/finishing steps with bed verification – doesn’t really matter, whichever makes most sense to the user. We did rename one button in LCM Triomics to make more sense.
* Looking at user interface mockups –
    * inline panel vs page – think users won’t mind – whichever is easier for us
    * think about whether need to filter buttons based on state, to make them show in relevant order
    * think also allowing creation of plates one by one would be confusing, if creating them all at once is an option
    * shouldn’t be able to create the ‘Denat’ plate from the lysate plate – should be an option on the final plate in the group
    * concerns about needing to choose multiple printers for 384, 96 and tube – that would be very niche – just 96 is fine for now
